docs/en/04_Changelogs/4.0.0.md

```diff
@@ -1095,6 +1095,55 @@ Some methods on `Requirements` have had their method signatures changed:
 A new config `Requirements_Backend.combine_in_dev` has been added in order to allow combined files to be
 forced on during development. If this is off, combined files is only enabled in live environments.
 
+Form validation has been refactored significantly. A new `FormMessage` trait has been created to
+handle field-level and form-level messages. This has the following properties:
+
+* `setMessage` to assign a message, type, and cast
+* `getMessage` retrieves the message string
+* `getMessageType` retrieves the message type (E.g. error, good, info)
+* `getMessageCast` retrieves the cast type
+* `getMessageCastingHelper` retrieves the DBField cast to use for the appropriate message cast
+* `getSchemaMessage` encodes this message for form schema use in ReactJS.
+
+`Form` methods have been changed:
+
+* `validate` method now returns a `ValidationResult` instance. This is no longer automatically
+  persisted in the state by default, unless a redirection occurs. You can also save any response in the
+  state by manually invoking `saveFormState` inside a custom validation response handler.
+* `setupFormErrors` renamed to `restoreFormState`
+* `resetValidation` renamed to `clearFormState`
+* `loadMessagesFrom` method created to load a ValidationResult into a form.
+* `setMessage`. third parameter is now $cast type
+* `messageForForm`. removed
+* `getSessionValidationResult` / `setSessionValidationResult` used to get / set session errors
+* `getSessionData` / `setSessionData` used to get / set field values cached in the session
+* `getAjaxErrorResponse` and `getRedirectReferer` created to simplify `getValidationErrorResponse`
+* `addErrorMessage` removed. Users can eithenr use `sessionMessage` or `sessionError` to add a
+  form level message, throw a ValidationException during submission, or add a custom validator.
+
+`Validator` methods have changed:
+
+* `validate` method now returns a `ValidationResult` instance.
+* `requireField` method removed. Use `RequiredFields` subclass instead.
+
+`ValidationResult` now has these methods:
+
+* `serialize` / `unserialize` for saving within session state
+* `messageList` renamed to `getMessages`
+* `getFieldMessages` created as a subset of `getMessages` for fields only.
+* `getOverallMessages` created as a subset of `getMessages` for non-fields only.
+* `codeList` renamed to `getMessageCodes`
+* `message` renamed to `joinedMessage`
+* `joinedOverallMessage` added.
+* `starredList` renamed to `starredMessage`
```

Comment:
Agreed. It's a hangover from the kitchen-sink mentality that SilverStripe is moving away from.